File: YNABInstantTransactions.Android/SmsReceiver.cs
using Android.App;
using Android.Content;
using Android.Provider;
using Android.Widget;
//using static YNABInstantTransaction.Domain.DomainTypes;
//using static YNABInstantTransaction.Domain.API;

namespace YNABInstantTransactions.Android
{
    [BroadcastReceiver(Name = "com.willettEnterprise.YNABInstantTransaction.SmsReceiver", Enabled = true, Exported = true)]
    [IntentFilter(new[] {"android.provider.Telephony.SMS_RECEIVED" }, Priority = (int)IntentFilterPriority.HighPriority)]
    public class SmsReceiver : BroadcastReceiver
    {
    //    private readonly API API = createAPI();
        public override void OnReceive(Context context, Intent intent)
        {
            System.Console.WriteLine("Got a text message!");

            foreach (var sms in Telephony.Sms.Intents.GetMessagesFromIntent(intent))
            {
                System.Console.WriteLine("Got message: " + sms.MessageBody + "; From: " + sms.OriginatingAddress);
                //          var result = API.ProcessSMS(sms.OriginatingAddress, sms.DisplayMessageBody);
//
 //               if (Microsoft.FSharp.Core.FSharpOption<Transaction>.get_IsSome(result))
  //              {
   //                 // @TODO - save to a file and send broadcast to main activity
    //                Toast.MakeText(context, "Got a transaction!", ToastLength.Short);
               // }
            }
        }
    }
}

File: YNABInstantTransactions.Android/SmsReceiver.fs
namespace YNABInstantTransactions.Android

open System

open Android
open Android.App
open Android.Support.V4.Content
open Android.Content
open Android.Provider

open YNABInstantTransactions
open YNABInstantTransactions.DataGateways

[<BroadcastReceiver(Name = "com.willettEnterprise.YNABInstantTransaction.SmsReceiver", Enabled = true, Exported = true)>]
[<IntentFilter([|"android.provider.Telephony.SMS_RECEIVED"|])>]
type SmsReceiver () =
    inherit BroadcastReceiver()

    override this.OnReceive (context, intent) =

        let broadcastTransaction (transactionID:Guid) =
            use smsIntent = new Intent("TRANSACTION_RECIEVED")
            let smsIntent = smsIntent.PutExtra("TransactionID", transactionID.ToString())
            LocalBroadcastManager.GetInstance(context).SendBroadcast(smsIntent) |> ignore

        let processSMS = UseCases.ProcessSMS.execute TransactionDataGateway.saveTransaction broadcastTransaction

        Telephony.Sms.Intents.GetMessagesFromIntent(intent) 
        |> Array.toSeq
        |> Seq.map (fun sms -> sms.DisplayMessageBody)
        |> Seq.iter processSMS
